platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/deduplication/EventDeduplicator.java

```diff
@@ -0,0 +1,137 @@
+
+    /**
+     * Deduplicated events are passed to this consumer.
+     */
+    private final Consumer<GossipEvent> eventConsumer;
+
+    /**
+     * The current minimum generation required for an event to be non-ancient.
+     */
+    private long minimumGenerationNonAncient = 0;
+
+    private static final LongAccumulator.Config DUPLICATE_EVENT_CONFIG = new LongAccumulator.Config(
+                    PLATFORM_CATEGORY, "duplicateEvents")
+            .withDescription("Events received that exactly match a previous event")
+            .withUnit("events");
+    private final LongAccumulator duplicateEventAccumulator;
+
+    private static final LongAccumulator.Config DISPARATE_SIGNATURE_CONFIG = new LongAccumulator.Config(
+                    PLATFORM_CATEGORY, "eventsWithDisparateSignature")
+            .withDescription(
+                    "Events received that match a descriptor of a previous event, but with a different signature")
+            .withUnit("events");
+    private final LongAccumulator disparateSignatureAccumulator;
+
+    /**
+     * A map from event descriptor to a set of signatures that have been received for that event.
+     */
+    private final SequenceMap<EventDescriptor, Set<byte[]>> observedEvents =
+            new StandardSequenceMap<>(0, INITIAL_CAPACITY, true, EventDescriptor::getGeneration);
+
+    /**
+     * Constructor
+     *
+     * @param platformContext the platform context
+     * @param eventConsumer   deduplicated events are passed to this consumer
+     */
+    public EventDeduplicator(
+            @NonNull final PlatformContext platformContext, @NonNull final Consumer<GossipEvent> eventConsumer) {
+
+        this.eventConsumer = Objects.requireNonNull(eventConsumer);
+
+        this.duplicateEventAccumulator = platformContext.getMetrics().getOrCreate(DUPLICATE_EVENT_CONFIG);
+        this.disparateSignatureAccumulator = platformContext.getMetrics().getOrCreate(DISPARATE_SIGNATURE_CONFIG);
+    }
+
+    /**
+     * Handle a potentially duplicate event
+     * <p>
+     * Ancient events are ignored. If the input event has not already been observed by this deduplicator, it is passed
+     * to the event consumer.
+     *
+     * @param event the event to handle
+     */
+    public void handleEvent(@NonNull final GossipEvent event) {
+        if (event.getGeneration() < minimumGenerationNonAncient) {
+            // Ancient events can be safely ignored.
+            return;
+        }
+
+        final EventDescriptor eventDescriptor = event.getDescriptor();
+
+        // a duplicate descriptor alone isn't sufficient to constitute a duplicate event
+        // we still need to check the signature
+        final boolean duplicateDescriptor = observedEvents.containsKey(eventDescriptor);
+
+        if (!duplicateDescriptor) {
+            observedEvents.put(eventDescriptor, new HashSet<>());
+        }
+
+        if (observedEvents.get(eventDescriptor).add(event.getUnhashedData().getSignature())) {
```

Comment:
Define this constant (to avoid the creation of a bunch of lambdas):

```
private static final Function<EventDescriptor, Set<byte[]>> BUILD_NEW_HASH_SET = k -> new HashSet<>();
```

Then, you can do the following:

```
final Set<byte[]> signatures = observedEvents.computeIfAbsent(eventDescriptor, BUILD_NEW_HASH_SET);

if (signatures.isEmpty() {
   // event descriptor is unique
} else {
    if (signatures.add(event.getUnhashedData().getSignature()) {
       // signature is unique, descriptor is not
    } else {
       // duplicate descriptor and signature
    }
}
```

This ends up only using a single map operation, as opposed to 3.